### PR TITLE
Maintenance Events resource was transfered to the APIv2

### PR DESCRIPTION
**Pull request by testisnullus**
Blocked until the https://instaclustr.atlassian.net/browse/INS-25547 issue is resolved

### CODE REVIEW

File: controllers/clusters/cadence_controller.go [hunk 1]
```diff
@@ -1241,3 +1221,28 @@ func (r *CadenceReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+
+	iMEStatuses, err := fetchMaintenanceEventStatuses(r.API, c.Status.ID)
+	if err != nil {
+		return err
+	}
+
+	if !c.Status.AreMaintenanceEventStatusesEqual(iMEStatuses) {
+		patch := c.NewPatch()
+		c.Status.MaintenanceEvents = iMEStatuses
+		err = r.Status().Patch(context.TODO(), c, patch)
```
Comment (ribaraka) [inline controllers/clusters/cadence_controller.go:1236]: please use ctx from the func param

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: controllers/clusters/helpers.go [hunk 2]
```diff
@@ -159,6 +161,38 @@ func getSortedAppVersions(versions []*models.AppVersions, appType string) []*ver
 	return nil
 }
 
+func fetchMaintenanceEventStatuses(
+	api instaclustr.API,
+	clusterID string,
+) ([]*clusterresourcesv1beta1.ClusteredMaintenanceEventStatus,
+	error) {
+
```
Comment (ribaraka) [inline controllers/clusters/helpers.go:169]: How about making it as a method for `instaclustr.API` client?

File: apis/clusterresources/v1beta1/maintenanceevents_types.go [hunk 3]
```diff
@@ -17,27 +17,24 @@ limitations under the License.
...
-	ClusterID                    string                            `json:"clusterId"`
-	MaintenanceEventsReschedules []*MaintenanceEventRescheduleSpec `json:"maintenanceEventsReschedule,omitempty"`
+	ClusterID                    string                        `json:"clusterId"`
+	MaintenanceEventsReschedules []*MaintenanceEventReschedule `json:"maintenanceEventsReschedule"`
 }
 
 // MaintenanceEventsStatus defines the observed state of MaintenanceEvents
 type MaintenanceEventsStatus struct {
-	EventsStatuses []*MaintenanceEventStatus `json:"eventsStatuses,omitempty"`
+	RescheduledEvent MaintenanceEventReschedule `json:"rescheduled"`
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply apis/clusterresources/v1beta1/maintenanceevents_types.go:32]: fixed

File: apis/clusterresources/v1beta1/maintenanceevents_types.go [hunk 4]
```diff
@@ -17,27 +17,24 @@ limitations under the License.
...
+}
+
+type MaintenanceEventReschedule struct {
+	ScheduledStartTime string `json:"scheduledStartTime"`
+	MaintenanceEventId string `json:"maintenanceEventId"`
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply apis/clusterresources/v1beta1/maintenanceevents_types.go:37]: fixed

File: apis/clusterresources/v1beta1/maintenanceevents_webhook.go [hunk 5]
```diff
@@ -71,3 +73,13 @@ func (r *MaintenanceEvents) ValidateDelete() error {
...
 	return nil
 }
+
+func (r *MaintenanceEvents) ValidateMaintenanceEventsReschedules() error {
+	for _, event := range r.Spec.MaintenanceEventsReschedules {
+		if dateValid, err := validation.ValidateISODate(event.ScheduledStartTime); err != nil || !dateValid {
+			return fmt.Errorf("scheduledStartTime must be provided in an ISO-8601 formatted UTC string: %v", err)
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply apis/clusterresources/v1beta1/maintenanceevents_webhook.go:83]: If the customer provides `scheduledStartTime` in an invalid format, we will return the error message

File: apis/clusters/v1beta1/structs.go [hunk 6]
```diff
@@ -509,6 +542,28 @@ func (c *Cluster) CloudProviderSettingsFromInstAPI(iDC models.DataCentre) (setti
 	return
 }
 
+func (c *Cluster) CloudProviderSettingsFromInstAPIv1(iProviders []*models.ClusterProviderV1) (accountName string, settings []*CloudProviderSettings) {
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply apis/clusters/v1beta1/structs.go:545]: Removed, it seems that I've accidentally added this method when resolving conflicts

File: config/samples/clusters_v1beta1_cassandra.yaml [hunk 7]
```diff
@@ -3,7 +3,7 @@ kind: Cassandra
 metadata:
   name: cassandra-cluster
 spec:
-  name: "username-Cassandra"
+  name: "danylo-Cassandra"
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply config/samples/clusters_v1beta1_cassandra.yaml:6]: fixed

File: controllers/clusters/helpers.go [hunk 8]
```diff
@@ -19,6 +19,8 @@ package clusters
 import (
 	"encoding/json"
 	"fmt"
+	clusterresourcesv1beta1 "github.com/instaclustr/operator/apis/clusterresources/v1beta1"
+	"github.com/instaclustr/operator/pkg/instaclustr"
 	"sort"
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply controllers/clusters/helpers.go:24]: fixed

File: controllers/clusters/cadence_controller.go (hunk 1 above)
Comment (testisnullus) [reply controllers/clusters/cadence_controller.go:1236]: fixed for all similar cases

File: controllers/clusters/helpers.go (hunk 2 above)
Comment (testisnullus) [reply controllers/clusters/helpers.go:169]: That sounds pretty logical, will move to the client

File: apis/clusterresources/v1beta1/maintenanceevents_types.go [hunk 9]
```diff
@@ -47,7 +44,16 @@ type MaintenanceEventStatus struct {
...
+	IsFinalized           bool   `json:"isFinalized"`
+	StartTime             string `json:"startTime,omitempty"`
+	EndTime               string `json:"endTime,omitempty"`
+	Outcome               string `json:"outcome,omitempty"`
+}
+
+type ClusteredMaintenanceEventStatus struct {
+	InProgress []*MaintenanceEventStatus `json:"inProgress"`
+	Past       []*MaintenanceEventStatus `json:"past"`
+	Upcoming   []*MaintenanceEventStatus `json:"Upcoming"`
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply apis/clusterresources/v1beta1/maintenanceevents_types.go:56]: will fix

File: controllers/clusters/cassandra_controller.go [hunk 10]
```diff
@@ -1224,6 +1204,31 @@ func (r *CassandraReconciler) deleteBackups(ctx context.Context, clusterID, name
...
+
+	iMEStatuses, err := fetchMaintenanceEventStatuses(r.API, c.Status.ID)
+	if err != nil {
+		return err
+	}
+
+	if !c.Status.AreMaintenanceEventStatusesEqual(iMEStatuses) {
+		patch := c.NewPatch()
+		c.Status.MaintenanceEvents = iMEStatuses
+		err = r.Status().Patch(context.TODO(), c, patch)
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply controllers/clusters/cassandra_controller.go:1218]: fixed

File: apis/clusters/v1beta1/structs.go [hunk 11]
```diff
@@ -419,26 +416,62 @@ func (c *Cluster) newImmutableFields() immutableCluster {
...
+	for i := range a.Upcoming {
+		if !areClusteredMaintenanceEventStatusEqual(a.Upcoming[i], b.Upcoming[i]) {
+			return false
+		}
+	}
 	return true
 }
 
+func areClusteredMaintenanceEventStatusEqual(a, b *clusterresource.MaintenanceEventStatus) bool {
+	return a.ID == b.ID &&
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply apis/clusters/v1beta1/structs.go]: Good catch, will fix

File: controllers/clusterresources/maintenanceevents_controller.go [hunk 12]
```diff
@@ -69,204 +67,114 @@ func (r *MaintenanceEventsReconciler) Reconcile(ctx context.Context, req ctrl.Re
...
+			return models.ReconcileRequeue, nil
 		}
 
-		var updated bool
-		patch := me.NewPatch()
-		instMEventsStatuses, err := r.API.GetMaintenanceEventsStatuses(me.Spec.ClusterID)
+		patch = me.NewPatch()
+		me.Status.RescheduledEvent.MaintenanceEventId = ""
+		me.Status.RescheduledEvent.ScheduledStartTime = ""
+		err = r.Status().Patch(ctx, me, patch)
```
[the remark this answers is not in the document]
Comment (testisnullus) [reply controllers/clusterresources/maintenanceevents_controller.go]: Valid point, will fix

**Review by worryg0d — APPROVED** [review]

**Review by OleksiienkoMykyta — APPROVED** [review]

File: apis/clusterresources/v1beta1/maintenanceevents_webhook.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ribaraka) [reply apis/clusterresources/v1beta1/maintenanceevents_webhook.go:83]: why then check for `!dateValid`, if there is always an error returned by invalid format?

**Review by ribaraka — APPROVED** [review]

**Pull request merged**